app/models/r_package.rb

```diff
@@ -1,53 +1,85 @@
 # Only module that should communicate with screensmart-r.
 module RPackage
-  def self.question_by_key(key)
-    questions.detect { |question| question['key'] == key }
+  def self.question_by_id(id)
+    questions.detect { |question| question['id'] == id }
   end
 
-  def self.question_keys
-    questions.map { |question| question['key'] }
+  def self.question_ids
+    questions.map { |question| question['id'] }
   end
 
   def self.questions
+    database['questions']
+  end
+
+  def self.domain_ids
+    domains.map { |domain| domain['id'] }
+  end
+
+  def self.domains
+    database['domains']
+  end
+
+  def self.database
     call('get_itembank_rdata')
   end
 
-  def self.data_for(answers)
-    raw_data = call('call_shadowcat', responses: [])
-    memo = rewrite_response_hash(raw_data)
+  # Retrieve a hash of attributes defined by the R package
+  # for a given set of answers (e.g. 'EL02' => 1)
+  # and domain_ids(e.g. ['POS-PQ'])
+  def self.data_for(answers, domain_ids)
+    raise 'No domains given' unless domain_ids.present?
 
-    answers.each_with_index do |_, index|
-      params = { responses: [answers.take(index + 1).to_h],
-                 estimate: memo[:estimate].try(:to_f),
-                 variance: memo[:variance].try(:to_f) }.compact
+    hash = normalized_shadowcat answers: [], domain: domain_ids
 
-      raw_data = call('call_shadowcat', params)
+    # Recalculate the estimate and variance for <index> answers,
+    # then recalculate it for <index + 1> answers with the previous estimate and variance,
+    # repeat until done for all answers
+    answers.each_with_index do |_, index|
```

Comment:
def self.data_for(answers, domain_ids)
    raise 'No domains given' unless domain_ids.present?

    hash = normalized_shadowcat answers: [], domain: domain_ids

    # Recalculate the estimate and variance for <index> answers,
    # then recalculate it for <index + 1> answers with the previous estimate and variance,
    # repeat until done for all answers
    answers.each_with_index.inject(hash) do |hash, (_, index)|
      normalized_shadowcat answers: [answers.take(index + 1).to_h],
                           estimate: hash[:estimate].try(:to_f),
                           variance: hash[:variance].try(:to_f),
                           domain: domain_ids
    end
  end